### DISCUSSION TITLE
Zen3/Vermeer CPPC

### DISCUSSION THREAD

**Discussion by gel-crabs** [cyring/CoreFreq#376] [General]
Hello, I have a Ryzen 5 5600X and I am unable to enable CPPC (CPPC/CPPC-PC is enabled in BIOS). acpi_cppc gives the preferred cores information however. I've gathered that the MSRs are not known for this model of CPU; is there any way I can help? I would be glad to provide any information.

**Comment by cyring** [comment]
> ... is there any way I can help? I would be glad to provide any information.

Sure, I will need the following output to verify how CPPC has been discovered.

```
corefreq-cli -s -n -m -n -k -n -B -n -M
```

**Reply by gel-crabs** [reply]
[corefreq.txt](https://github.com/cyring/CoreFreq/files/10059206/corefreq.txt)

**Reply by cyring** [reply]
Is your kernel built with ACPI CPPC support ?
You should find some CPPC directives in the `/proc/config*`

**Reply by gel-crabs** [reply]
I believe so.
[config.txt](https://github.com/cyring/CoreFreq/files/10059885/config.txt)

**Reply by cyring** [reply]
# Prerequisites
```
CONFIG_ACPI_CPPC_LIB=y
```
# Debugging
I have instrumented driver code with some `pr_warn` and `pr_info`
and `"CoreFreq: 'acpi_cpc' Execution failure code 1"` is showing up in my config case:
1. BIOS CPPC is enabled
2. ACPI `_CPC` table is enabled
3. `x2APIC` is set rather than `AUTO`
  _which, I have noticed, is disabling ACPI CPPC, the 2 previous rules._
```C
signed int Read_ACPI_CPPC_Registers(unsigned int cpu, void *arg)
{
#ifdef CONFIG_ACPI_CPPC_LIB
	struct cppc_perf_fb_ctrs CPPC_Perf;
	struct cppc_perf_caps CPPC_Caps;

	CORE_RO *Core = (CORE_RO *) PUBLIC(RO(Core, AT(cpu)));

	signed int rc = 0;
	if ((rc = cppc_get_perf_ctrs(Core->Bind, &CPPC_Perf)) == 0) {
		rc = cppc_get_perf_caps(Core->Bind, &CPPC_Caps);
	}
	if (rc == 0) {
	    #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 1, 0)
		unsigned long long desired_perf = 0;
	    #endif
		Core->PowerThermal.ACPI_CPPC = (struct ACPI_CPPC_STRUCT) {
			.Highest	= CPPC_Caps.highest_perf,
			.Guaranteed	= CPPC_Caps.nominal_perf,
			#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 18, 0)
			.Efficient	= CPPC_Caps.nominal_freq,
			.Lowest 	= CPPC_Caps.lowest_freq,
			.Minimum	= CPPC_Caps.lowest_freq,
			#else
			.Efficient	= CPPC_Caps.nominal_perf,
			.Lowest 	= CPPC_Caps.lowest_perf,
			.Minimum	= CPPC_Caps.lowest_perf,
			#endif
			.Maximum	= CPPC_Caps.highest_perf,
			.Desired	= CPPC_Perf.reference_perf,
			.Energy 	= 0
		};
	    #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 1, 0)
		rc = cppc_get_desired_perf(Core->Bind, &desired_perf);
		if (rc == 0) {
			Core->PowerThermal.ACPI_CPPC.Desired = desired_perf;
		}
	    #endif
	} else {
		pr_warn("CoreFreq: 'cppc' Execution failure code %d\n", rc);
	}
	return rc;
#else
	pr_warn("CoreFreq: No CONFIG_ACPI_CPPC_LIB support\n");
	return -ENODEV;
#endif /* CONFIG_ACPI_CPPC_LIB */
}

void For_All_ACPI_CPPC(signed int(*CPPC_Func)(unsigned int, void*), void *arg)
{
	#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 11, 0)
	signed int rc = acpi_cpc_valid() == false;
	#else
	signed int rc = acpi_disabled;
	#endif

	unsigned int cpu;
	pr_info("CoreFreq: 'acpi_cpc' Execution failure code %d\n", rc);
	for (cpu = 0; (cpu < PUBLIC(RO(Proc))->CPU.Count) && (rc == 0); cpu++)
	{
		rc = CPPC_Func(cpu, arg);
	}
	PUBLIC(RO(Proc))->Features.ACPI_CPPC = (rc == 0);
}
```
* Your kernel version being greater than `5.11.0`, I believe the [`acpi_cpc_valid()`](https://elixir.bootlin.com/linux/latest/source/drivers/acpi/cppc_acpi.c#L422) call is also failing.
  `acpi_cpc_valid` is also being employed by kernel within other drivers as seen in this [symbol search](https://elixir.bootlin.com/linux/latest/A/ident/acpi_cpc_valid):
```
drivers/acpi/cppc_acpi.c, line 435
drivers/base/arch_topology.c, line 356
drivers/cpufreq/amd-pstate.c, line 685
drivers/cpufreq/cppc_cpufreq.c, line 950
```

_EDIT_: I have rebooting with the passing case: 
  In BIOS, local APIC is set to `AUTO`.  
  See section BIOS tab `CBS > CPU`

ACPI CPPC firmware is now implementing and giving service.
This also confirmed by `modprobe amd_pstate` which is successfully inserted this module and _CoreFreq_ is able to retrieve CPPC data.

![2022-11-22-075351_644x550_scrot](https://user-images.githubusercontent.com/11563789/203245209-bf7ce94e-f207-4ef1-b902-165931962fb1.png)

**Comment by cyring** [comment]
Using commit deee6015205891a470a1c25ca5c1e3258b4a5de0 we will now track the BIOS support of `_CPC` as retrieved by the kernel interface:

# _CPC disabled

![_CPC-BIOS_Disabled](https://user-images.githubusercontent.com/11563789/203256135-19107c02-6789-4753-a708-775d80ac82f2.png)

![2022-11-22-084916_644x550_scrot](https://user-images.githubusercontent.com/11563789/203256251-e93b79e7-b606-4321-bc1b-eceea5db5c47.png)

# _CPC enabled but another feature ...

![221122085400](https://user-images.githubusercontent.com/11563789/203257682-4610fca3-4024-4a30-ab9c-77ed88ac25d7.png)
![221122085422](https://user-images.githubusercontent.com/11563789/203257718-a48a6aec-9583-4476-bb69-23034e49e3e5.png)
![221122085439](https://user-images.githubusercontent.com/11563789/203257739-8a64c003-0af7-4175-a326-4a7f2fe75172.png)

![2022-11-22-090003_644x550_scrot](https://user-images.githubusercontent.com/11563789/203257862-68aadd9b-5793-44ba-b713-be526750d85d.png)

# _CPC enabled without any feature preventing Firmware

![221122090222](https://user-images.githubusercontent.com/11563789/203259200-9ee90bca-33f7-43cf-b43e-42c62f235bd3.png)
![221122090245](https://user-images.githubusercontent.com/11563789/203259247-9728f205-0ea5-438a-8337-017fb440aea9.png)
![221122090305](https://user-images.githubusercontent.com/11563789/203259278-67815275-d19f-4f0e-8d6d-265ef0ecdf16.png)

![2022-11-22-090430_644x550_scrot](https://user-images.githubusercontent.com/11563789/203259317-d1c61732-2fba-492b-9f6c-11900c2396d8.png)

**Comment by cyring** [comment]
# Zen2/Lucienne
This `5300U` is apparently gifted with **hardware** CPPC
_EDIT_ hardware state fixed in CLI of commit 04eb58c8896efbba03adf22f5043d1c2447ca461

## Initial state as disabled
_thus _CPC does not matter in this case_
![2022-11-22-095823_644x424_scrot](https://user-images.githubusercontent.com/11563789/203270413-4f5681a2-0c6d-4e9b-80a3-08a3fad314a7.png)

## Request to enable CPPC
![2022-11-22-095855_644x424_scrot](https://user-images.githubusercontent.com/11563789/203270501-6c10ffa1-cb16-49c8-80d8-e289611feef5.png)

## CPPC activated
![2022-11-22-095939_644x424_scrot](https://user-images.githubusercontent.com/11563789/203270680-16e15798-c8a1-4263-a40f-418ee21a481e.png)

**Comment by gel-crabs** [comment]
Holy crap, god bless you. My BIOS does not have any setting for X2APIC/Local APIC, sadly ;-; (ASRock B450M Steel Legend)

Do you know of any other names for this setting?

**Reply by cyring** [reply]
Can you please pull, build and run the **latest** branch [`develop`](https://github.com/cyring/CoreFreq/tree/develop), next show me a screenshot of _CoreFreq_ UI with the `Performance Monitoring` window opened.
I would like to check your current state of `_CPC`.

**Reply by gel-crabs** [reply]
![Screenshot_20221122_104606](https://user-images.githubusercontent.com/43506670/203358374-b92841b0-2bab-4db2-bcb5-0394b7a3e689.png)

Always running on develop.

**Reply by cyring** [reply]
Oh ! I was not expecting this result.
So we have `_CPC` but not the renaming support for `CPPC` which means driver reaches this line of code:
https://github.com/cyring/CoreFreq/blob/04eb58c8896efbba03adf22f5043d1c2447ca461/corefreqk.c#L3884

`Read_ACPI_CPPC_Registers` is indirectly the function called by the loop:
https://github.com/cyring/CoreFreq/blob/04eb58c8896efbba03adf22f5043d1c2447ca461/corefreqk.c#L3828

but `rc` is leaving function with a failure due to `cppc_get_perf_ctrs` or `cppc_get_perf_caps` or `cppc_get_desired_perf`

Sorry for the debug session but I have to examine again those 3 kernel functions for the reason of their failure.

**Reply by gel-crabs** [reply]
> Oh ! I was not expecting this result. So we have `_CPC` but not the renaming support for `CPPC` which means driver reaches this line of code:
> 
> https://github.com/cyring/CoreFreq/blob/04eb58c8896efbba03adf22f5043d1c2447ca461/corefreqk.c#L3884
> 
> `Read_ACPI_CPPC_Registers` is indirectly the function called by the loop:
> 
> https://github.com/cyring/CoreFreq/blob/04eb58c8896efbba03adf22f5043d1c2447ca461/corefreqk.c#L3828
> 
> but `rc` is leaving function with a failure due to `cppc_get_perf_ctrs` or `cppc_get_perf_caps` or `cppc_get_desired_perf`
> 
> Sorry for the debug session but I have to examine again those 3 kernel functions for the reason of their failure.

Do you need me to debug anything?

**Reply by cyring** [reply]
Yes please.

1. at this function `Read_ACPI_CPPC_Registers`
https://github.com/cyring/CoreFreq/blob/04eb58c8896efbba03adf22f5043d1c2447ca461/corefreqk.c#L3828
2. replace with the code below.

```c
signed int Read_ACPI_CPPC_Registers(unsigned int cpu, void *arg)
{
#ifdef CONFIG_ACPI_CPPC_LIB
	struct cppc_perf_fb_ctrs CPPC_Perf;
	struct cppc_perf_caps CPPC_Caps;

	CORE_RO *Core = (CORE_RO *) PUBLIC(RO(Core, AT(cpu)));

	signed int rc = 0;
	if ((rc = cppc_get_perf_ctrs(Core->Bind, &CPPC_Perf)) == 0) {
	    if ((rc = cppc_get_perf_caps(Core->Bind, &CPPC_Caps)) != 0)
		pr_err("CoreFreq: cppc_get_perf_caps() error %d\n", rc);
	} else {
		pr_err("CoreFreq: cppc_get_perf_ctrs() error %d\n", rc);
	}
	if (rc == 0) {
	    #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 1, 0)
		unsigned long long desired_perf = 0;
	    #endif
		Core->PowerThermal.ACPI_CPPC = (struct ACPI_CPPC_STRUCT) {
			.Highest	= CPPC_Caps.highest_perf,
			.Guaranteed	= CPPC_Caps.nominal_perf,
			#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 18, 0)
			.Efficient	= CPPC_Caps.nominal_freq,
			.Lowest 	= CPPC_Caps.lowest_freq,
			.Minimum	= CPPC_Caps.lowest_freq,
			#else
			.Efficient	= CPPC_Caps.nominal_perf,
			.Lowest 	= CPPC_Caps.lowest_perf,
			.Minimum	= CPPC_Caps.lowest_perf,
			#endif
			.Maximum	= CPPC_Caps.highest_perf,
			.Desired	= CPPC_Perf.reference_perf,
			.Energy 	= 0
		};
	    #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 1, 0)
		rc = cppc_get_desired_perf(Core->Bind, &desired_perf);
		if (rc == 0) {
			Core->PowerThermal.ACPI_CPPC.Desired = desired_perf;
		} else {
		    pr_err("CoreFreq: cppc_get_desired_perf() error %d\n", rc);
		}
	    #endif
	}
	return rc;
#else
	return -ENODEV;
#endif /* CONFIG_ACPI_CPPC_LIB */
}
```

3. rebuild and reload driver 
4. then search within the kernel log for any error trace like one of those:
```
CoreFreq: cppc_get_perf_caps() error 
CoreFreq: cppc_get_perf_ctrs() error
CoreFreq: cppc_get_desired_perf() error
```

**Reply by gel-crabs** [reply]
![Screenshot_20221122_114251](https://user-images.githubusercontent.com/43506670/203371705-685a78be-4a8f-417f-a115-ff8b02c82403.png)

Yep, here's the error.

![Screenshot_20221122_114331](https://user-images.githubusercontent.com/43506670/203371818-f0c50121-330c-4cec-a7b6-dc6e748bf4f6.png)

(My performance monitoring window)

**Reply by cyring** [reply]
`CoreFreq: cppc_get_desired_perf() error -22` means function call failed with [`EINVAL		22	/* Invalid argument */`](https://elixir.bootlin.com/linux/latest/source/include/uapi/asm-generic/errno-base.h#L26)

Call flow should have ended in kernel function [cppc_get_perf()](https://elixir.bootlin.com/linux/latest/source/drivers/acpi/cppc_acpi.c#L1089) but I don't find why an EINVAL exit happened.

**Reply by gel-crabs** [reply]
> `CoreFreq: cppc_get_desired_perf() error -22` means function call failed with [`EINVAL 22 /* Invalid argument */`](https://elixir.bootlin.com/linux/latest/source/include/uapi/asm-generic/errno-base.h#L26)
> 
> Call flow should have ended in kernel function [cppc_get_perf()](https://elixir.bootlin.com/linux/latest/source/drivers/acpi/cppc_acpi.c#L1089) but I don't find why an EINVAL exit happened.

Do you want more info? I believe I have ftrace/audit disabled.

**Reply by cyring** [reply]
I'm not sure if we have diverging kernel built but I'm developing on ArchLinux.

www.cyring.fr , link DEVELOPMENT, I'm providing an ISO image with ArchLinux + CoreFreq. It's an autorun image, you just have to `dd` on a USB key, boot it and let it go. 

If you have CPPC enabled then issue is within kernel. If not then ACPI CPPC could be out of specs inside Asrock BIOS implementation and kernel is failing at `_CPC` interface.

**Comment by gel-crabs** [comment]
I just tried the Live ISO and it works. It is just an issue with my kernel, I have linux-cachyos-bore with the amd_pstate_epp patches so that's the problem.

**Reply by gel-crabs** [reply]
Took the amd_pstate patch out of the kernel, recompiled, and now it works. God bless you cyring

**Reply by cyring** [reply]
Your welcome. 
Thank you for using _CoreFreq_

Feel free to update the CLI output and screenshots. I have added the on-going into the Wiki [CPU Support](https://github.com/cyring/CoreFreq/wiki/CPU-support), processor Ryzen 5 5600X

I was not aware of the [BORE](https://github.com/CachyOS/linux-cachyos) scheduler, but those look like major kernel patches. 

The same we, Contributers, did for MUQSS and BMQ; we can  add BORE to kernel flavors supported by _CoreFreq_
Here's the list of AUR packages found for ArchLinux:
https://aur.archlinux.org/packages?O=0&K=BORE&SB=p&SO=d
Let me know if you recommend one package among the others.

Regards

**Reply by gel-crabs** [reply]
> Your welcome. Thank you for using _CoreFreq_
> 
> Feel free to update the CLI output and screenshots. I have added the on-going into the Wiki [CPU Support](https://github.com/cyring/CoreFreq/wiki/CPU-support), processor Ryzen 5 5600X
> 
> I was not aware of the [BORE](https://github.com/CachyOS/linux-cachyos) scheduler, but those look like major kernel patches.
> 
> The same we, Contributers, did for MUQSS and BMQ; we can add BORE to kernel flavors supported by _CoreFreq_ Here's the list of AUR packages found for ArchLinux: https://aur.archlinux.org/packages?O=0&K=BORE&SB=p&SO=d Let me know if you recommend one package among the others.
> 
> Regards

I recommend linux-cachyos-bore, I've used it for a long time and it is the default CachyOS kernel.

Also, the amd_pstate patch I was referring to (amd_pstate_epp v4) is going to be merged into the kernel relatively soon, either 6.1 or 6.2, it's the newest version of the amd_pstate driver essentially. It is in linux-cachyos-bore by default for 6.0 and 6.1-rc though.

**Reply by cyring** [reply]
> > Your welcome. Thank you for using _CoreFreq_
> > Feel free to update the CLI output and screenshots. I have added the on-going into the Wiki [CPU Support](https://github.com/cyring/CoreFreq/wiki/CPU-support), processor Ryzen 5 5600X
> > I was not aware of the [BORE](https://github.com/CachyOS/linux-cachyos) scheduler, but those look like major kernel patches.
> > The same we, Contributers, did for MUQSS and BMQ; we can add BORE to kernel flavors supported by _CoreFreq_ Here's the list of AUR packages found for ArchLinux: https://aur.archlinux.org/packages?O=0&K=BORE&SB=p&SO=d Let me know if you recommend one package among the others.
> > Regards
> 
> I recommend linux-cachyos-bore, I've used it for a long time and it is the default CachyOS kernel.

I'm choosing it but `PKGBUILD` stops with:
```
==> Validating source files with sha256sums...
    linux-6.0.9.tar.xz ... Passed
    config ... Passed
    config-rt ... Passed
    auto-cpu-optimization.sh ... Passed
    0001-cachyos-base-all.patch ... Passed
    0001-Add-latency-priority-for-CFS-class.patch ... Passed
    0001-bore-cachy.patch ... FAILED
```
I have added the key `F3B607488DB35A47` mentionned in [README.md](https://github.com/CachyOS/linux-cachyos#readme) without success.
To go on, I'm altering the `sha256sums=` with a `SKIP` instead. Not my favorite solution.

> Also, the amd_pstate patch I was referring to (amd_pstate_epp v4) is going to be merged into the kernel relatively soon, either 6.1 or 6.2, it's the newest version of the amd_pstate driver essentially. It is in linux-cachyos-bore by default for 6.0 and 6.1-rc though.

It is clear now that I've to make _CoreFreq_ compatible with those patches.

I will come back here when I'm done experimenting BORE and the preliminary developments for the driver.

**Reply by gel-crabs** [reply]
Oh yeah, those can happen with linux-cachyos-bore because the patches get updated frequently. I just pass --skipinteg to makepkg as the patches are pulled from the CachyOS Github anyway.

Also, I started using CoreFreq on a Ryzen 5 2400G due to the TSC being unreliable, so I know for a fact CoreFreq works on that CPU.

**Comment by cyring** [comment]
`6.0.9-1-cachyos-bore` is booted on my 3950X and I can reproduce your issue.

Looking at kernel, Patch has changed code as below:
```c
static int cppc_get_perf(int cpunum, enum cppc_regs reg_idx, u64 *perf)
{
        struct cpc_desc *cpc_desc = per_cpu(cpc_desc_ptr, cpunum);
        struct cpc_register_resource *reg;
        int pcc_ss_id = per_cpu(cpu_pcc_subspace_idx, cpunum);
        struct cppc_pcc_data *pcc_ss_data = NULL;
        int ret = -EINVAL;

        if (!cpc_desc) {
                pr_debug("No CPC descriptor for CPU:%d\n", cpunum);
                return -ENODEV;
        }

        reg = &cpc_desc->cpc_regs[reg_idx];

        if (CPC_IN_PCC(reg)) {
                if (pcc_ss_id < 0)
                        return -EIO;

                pcc_ss_data = pcc_data[pcc_ss_id];

                down_write(&pcc_ss_data->pcc_lock);

                if (send_pcc_cmd(pcc_ss_id, CMD_READ) >= 0)
                        cpc_read(cpunum, reg, perf);
                else
                        ret = -EIO;

                up_write(&pcc_ss_data->pcc_lock);

                return ret;
        }

        cpc_read(cpunum, reg, perf);

        return ret;
}
```

`ret` is initialized with `-EINVAL` which is the `-22` error code we are getting from our trace.
In the original interface of the kernel [cppc_get_perf](https://elixir.bootlin.com/linux/latest/C/ident/cppc_get_perf) is supposed to return zero to say all's good.
But here Patch makes `ret` returns `-EINVAL` if no error happened !
Thus _CoreFreq_ catches this error but also `amd_pstate` which apparently says silently: no no I won't load.

I can change _CoreFreq_ to handle `-EINVAL` as a _good_ return code but `amd_pstate` will remain unsolved.
So an issue has to be opened to the Patch author (_or there's something I have not understood from this Patch_)

**Comment by cyring** [comment]
Based on your indication, I have search and found this patch
https://lore.kernel.org/lkml/DM4PR12MB5278668BD034D950AB7416AB9C099@DM4PR12MB5278.namprd12.prod.outlook.com/T/

`-EINVAL` is apparently part of the new logic but I still don't get if its handling by Callers is an issue or not.

If all theses changes are going into future mainstream 6.1 or later versions, it's worth working on CoreFreq driver adaptions when release candidates will show up. 

_EDIT: See below for the committed workaround_

**Comment by cyring** [comment]
Meanwhile I have made a working fix for `6.0.9-1-cachyos-bore`
![2022-11-23-051634_644x564_scrot](https://user-images.githubusercontent.com/11563789/203468410-69fda44b-70c3-4678-b599-a98cf9ee3ee3.png)
Now rebooting with Mainstream kernel to check for non regressions.
_EDIT_: Tests OK

**Comment by cyring** [comment]
Commit 6b80b8501a3ec1431855e3a11286b9801404d73d is adding the support of the build directive `CONFIG_SCHED_BORE` and accepts `-EINVAL` as a passing case.

You should now be able to run _CoreFreq_ with no change in your kernel setup.

**Reply by gel-crabs** [reply]
Wow. Thank you so much. Do you have a donation link?

Also, the CachyOS repos now have corefreq-git!

**Reply by cyring** [reply]
> Wow. Thank you so much. Do you have a donation link?

I'm still asking myself about that. _CoreFreq_ motivations are: my Engineer story ...
But I need Hardware and Programmers to go fast, life is short ...
Need to train volunteers to pursue the project ...

> Also, the CachyOS repos now have corefreq-git!

Love this. 
Is this available on Github or Web site ?

**Comment by cyring** [comment]
Please come back here if you are aware of any change in this CPPC lib implementation.
I'll try to keep _CoreFreq_ driver logic in sync with CONFIG_BORE

**Reply by gel-crabs** [reply]
Thank you!! BORE was just updated from 1.7.0 to 1.7.1 on linux-cachyos-bore, I just updated, corefreqk doesn't give any errors or anything so I assume everything's fine but just wanted to let you know.

https://github.com/CachyOS/kernel-patches/commit/cb044be44da65d22d8aaf867d63822f3d9016a8d

https://github.com/CachyOS
https://wiki.cachyos.org/en/home/Repo (x86_64-v3 compiled packages, updated frequently!)

**Reply by cyring** [reply]
Thank you.

I wanted to try `amd_pstate_ut` to check how its renders into _CoreFreq_ (item `CPU-Freq driver`)

But `amd_pstate_ut` refuses to load with these traces left in kernel log:
```
amd_pstate_ut: 1    amd_pstate_ut_acpi_cpc_valid         success!
amd_pstate_ut: get_shared_mem unable to open /sys/module/amd_pstate/parameters/shared_mem file!
amd_pstate_ut: amd_pstate_ut_pstate_enable amd pstate must be enabled!
amd_pstate_ut: 2    amd_pstate_ut_check_enabled         fail!
amd_pstate_ut: 3    amd_pstate_ut_check_perf         success!
amd_pstate_ut: 4    amd_pstate_ut_check_freq         success!
```

Based on its source code I've added `amd_pstate=legacy_cppc` in the kernel boot command line but it does not make any difference.

Fyi my full command line which works perfectly with the mainstream `amd_pstate` of kernel `6.0` versions

```
add_efi_memmap nmi_watchdog=0 selinux=0 loglevel=3 rd.systemd.show_status=auto rd.udev.log-priority=3 consoleblank=0 vt.color=0x03 modprobe.blacklist=joydev,mousedev,pcspkr,k10temp,sp5100_tco,rapl,intel_rapl_msr,intel_rapl_common,acpi_cpufreq cpu0_hotplug audit=0 nowatchdog mitigations=off nokaslr no_hash_pointers sysrq_always_enabled msr.allow_writes=on initcall_blacklist=acpi_cpufreq_init idle=halt tsc=unstable acpi_enforce_resources=lax amdgpu.ppfeaturemask=0xffffffff amd_pstate=legacy_cppc
```

I've also "fixed" the CPPC lib to return zero rather than `-EINVAL` but it still don't load.

Does `amd_pstate_ut` work for you ?

**Reply by cyring** [reply]
> Thank you!! BORE was just updated from 1.7.0 to 1.7.1 on linux-cachyos-bore, I just updated, corefreqk doesn't give any errors or anything so I assume everything's fine but just wanted to let you know.
> 
> [CachyOS/kernel-patches@cb044be](https://github.com/CachyOS/kernel-patches/commit/cb044be44da65d22d8aaf867d63822f3d9016a8d)
> 
> https://github.com/CachyOS https://wiki.cachyos.org/en/home/Repo (x86_64-v3 compiled packages, updated frequently!)

Can you make the patch available through the AUR package ?

**Reply by cyring** [reply]
> Thank you!! BORE was just updated from 1.7.0 to 1.7.1 on linux-cachyos-bore, I just updated, corefreqk doesn't give any errors or anything so I assume everything's fine but just wanted to let you know.
> 
> [CachyOS/kernel-patches@cb044be](https://github.com/CachyOS/kernel-patches/commit/cb044be44da65d22d8aaf867d63822f3d9016a8d)
> 
> https://github.com/CachyOS https://wiki.cachyos.org/en/home/Repo (x86_64-v3 compiled packages, updated frequently!)

Commit 4bf9f02d6fa20d144e3508e5b99bfabf59574cb0 within `develop` branch now builds with CachyOS patch to manage the CPPC/EPP.  

![2022-12-12-104612_644x1012_scrot](https://user-images.githubusercontent.com/11563789/207013689-37e29bee-2ac2-4f77-af99-1ffce46f090a.png)

I've altered the AUR `PKGBUILD` to force a `menuconfig` then set directive `CONFIG_CPU_FREQ_DEFAULT_GOV_POWERSAVE=y` rather than `CONFIG_CPU_FREQ_DEFAULT_GOV_PERFORMANCE`. 
Without such build config, _CoreFreq_ driver could not employ `cppc_get_perf_ctrs()` and `cppc_set_epp_perf()`

**Comment by cyring** [comment]
@gel-crabs Hello,
Sorry for all these questions, I did not know `amd_pstate_ut` is for testings and requires `amd_pstate`
> * The AMD P-State Unit Test is a test module for testing the amd-pstate
> * driver. 1) It can help all users to verify their processor support

Usually when a driver has dependencies, it has to be specified. But here, none.

**Reply by cyring** [reply]
I found driver `amd_pstate_ut` exists in `6.1` only

Now getting these results:
```
amd_pstate_ut: 1    amd_pstate_ut_acpi_cpc_valid         success!
amd_pstate_ut: 2    amd_pstate_ut_check_enabled         success!
amd_pstate_ut: 3    amd_pstate_ut_check_perf         success!
amd_pstate_ut: 4    amd_pstate_ut_check_freq         success!
```
_Not very detailed, isn't it._

----

What I find significant is within function [`amd_pstate_ut_check_perf`](https://elixir.bootlin.com/linux/v6.1-rc6/C/ident/amd_pstate_ut_check_perf): [`amd_get_highest_perf`](https://elixir.bootlin.com/linux/v6.1-rc6/C/ident/amd_get_highest_perf) informs us that the highest performance is split in two ranges: `166` and `255` since [`5.12.15`](https://elixir.bootlin.com/linux/v5.12.5/source/arch/x86/kernel/cpu/amd.c#L1174), depending on the Zen processor family.
Still experimenting it but it will probably impact the CPPC capability ratios.
**EDIT**
According to [`amd_get_highest_perf()`](https://elixir.bootlin.com/linux/v6.0.9/source/arch/x86/kernel/cpu/amd.c#L1180) kernel function, my Matisse 3950X of CPUID `8F_71` should get `166` as the highest performance but dumping values show values between `166` and `255` !
```
CPU#0   ACPI_CPPC.Highest=236
CPU#1   ACPI_CPPC.Highest=236
CPU#2   ACPI_CPPC.Highest=231
CPU#3   ACPI_CPPC.Highest=226
CPU#4   ACPI_CPPC.Highest=211
CPU#5   ACPI_CPPC.Highest=221
CPU#6   ACPI_CPPC.Highest=206
CPU#7   ACPI_CPPC.Highest=216
CPU#8   ACPI_CPPC.Highest=191
CPU#9   ACPI_CPPC.Highest=201
CPU#10  ACPI_CPPC.Highest=186
CPU#11  ACPI_CPPC.Highest=196
CPU#12  ACPI_CPPC.Highest=171
CPU#13  ACPI_CPPC.Highest=181
CPU#14  ACPI_CPPC.Highest=166
CPU#15  ACPI_CPPC.Highest=176
CPU#16  ACPI_CPPC.Highest=236
CPU#17  ACPI_CPPC.Highest=236
CPU#18  ACPI_CPPC.Highest=231
CPU#19  ACPI_CPPC.Highest=226
CPU#20  ACPI_CPPC.Highest=211
CPU#21  ACPI_CPPC.Highest=221
CPU#22  ACPI_CPPC.Highest=206
CPU#23  ACPI_CPPC.Highest=216
CPU#24  ACPI_CPPC.Highest=191
CPU#25  ACPI_CPPC.Highest=201
CPU#26  ACPI_CPPC.Highest=186
CPU#27  ACPI_CPPC.Highest=196
CPU#28  ACPI_CPPC.Highest=171
CPU#29  ACPI_CPPC.Highest=181
CPU#30  ACPI_CPPC.Highest=166
CPU#31  ACPI_CPPC.Highest=176
```

Here's the instruction to print values in _CoreFreq_:
```c
signed int Read_ACPI_CPPC_Registers(unsigned int cpu, void *arg)
{
...
	printk("CPU#%d\tACPI_CPPC.Highest=%d\n", cpu, CPPC_Caps.highest_perf);
	}
	return rc;
#else
	return -ENODEV;
#endif /* CONFIG_ACPI_CPPC_LIB */
}
```

**Comment by cyring** [comment]
Since commit 80d6e11caca8b4e6f2d1d134e94f16f9a148ff23 HWP or CPPC capabilities are refactored to print values per CPU

# Intel Hardware HWP
## Before enabling HWP
![Intel_HWP_OFF](https://user-images.githubusercontent.com/11563789/204637495-df839da5-2bb1-4fda-b22a-601cc44dbc48.png)
## HWP enabled
![Intel_HWP_ON](https://user-images.githubusercontent.com/11563789/204637564-c7a429bb-a92d-4f84-aa18-8892e25eca04.png)
### Remark (_EDIT_)
* `Efficient` is a varying element with this processor.
  Press `*` key to force refresh

# AMD Firmware CPPC
* As configured by driver `amd_pstate`

![2022-11-29-205516_644x1012_scrot](https://user-images.githubusercontent.com/11563789/204637832-fbfe86a1-cfb1-40d9-9655-b4701967a6d2.png)

## Remarks
* Capability performance values are scaled to a frequency **Ratio** value
  This is to have a common semantic between both manufacturers implementation.
* `Highest` is apparently the varying _element_ in my platform.
   I believe it reflects the so called _Best Core_
* `amd_pstate` is not mandatory for CPPC. 
   _CoreFreq_ can manage it as fully manual.
   User just has to enable/disable `CPPC       <FMW>` in `Performance Monitoring` window
   and then adjust `CPPC` `>` `TGT` in `Processor` window
   The same can be achieved from the driver using parameters `HWP_Enable` and an array of ratios in `Ratio_HWP`


Please try and let me know ...

**Comment by cyring** [comment]
* Commit 6794d6c971e7af519c68fc1cc81bb177913044b6 : _CoreFreq_ is maintaining the bounds of `Highest` capability whilst CPPC firmware is enabled / disabled

* Still an unsolved issue is the obtained operating frequency ratio after altering the `HWP-TGT` ratio from `Processor` window. (Prerequisite is to **enable** CPPC FMW in `Performance Monitoring` window)

  _Something inherent to firmware that I have no idea about._

  For example, `13` is entered here for all Cores but the obtained Cores frequency ratio is `18`
![2022-11-30-221519_644x1012_scrot](https://user-images.githubusercontent.com/11563789/204910920-5bc247bc-2584-4105-8d2f-0421388dd6bc.png)
![2022-11-30-221952_644x1012_scrot](https://user-images.githubusercontent.com/11563789/204910954-7d0aa2a9-b8d3-4b3a-83d8-59767fa45801.png)

  Remarks:
  - In the UI, you have to switch to `Absolute frequency`; press the `!` shortcut.
  - No `amd_pstate` involved here. This is a solely _CoreFreq_ control mode of CPPC.

**Comment by cyring** [comment]
TODO when [those](https://lore.kernel.org/all/20221128170314.2276636-2-perry.yuan@amd.com/) get merged into kernel:
```
cppc_get_epp_caps
cppc_set_epp_perf
```

Will impact this line:
https://github.com/cyring/CoreFreq/blob/6a4b450e48a735a8a67c169fdf9951583094266e/corefreqk.c#L3911

**Comment by cyring** [comment]
Since kernel `6.0.11` `amd-pstate` is forced built-in in Arch. 
So far the only working `cmdline` blacklist instructions are:

* Hardware CPPC (MSR registers)
`initcall_blacklist=amd_pstate_init`

* Firmware CPPC (ACPI registers)
`amd_pstate.shared_mem=0` and/or `initcall_blacklist=acpi_cpufreq_init`
